File: ReactAndroid/src/main/java/com/facebook/react/views/view/CanvasUtil.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.views.view;

import android.annotation.SuppressLint;
import android.graphics.Canvas;
import android.os.Build;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import javax.annotation.Nullable;

/**
 * Copied from <a
 * href="https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-graphics/src/androidMain/kotlin/androidx/compose/ui/graphics/CanvasUtils.android.kt;drc=3b2dde134afab8d58b9c39ad4820eaf9a6e014a9">
 * Compose canvas utils </a>
 */
public class CanvasUtil {
  private CanvasUtil() {}

  private @Nullable static Method mReorderBarrierMethod = null;
  private @Nullable static Method mInorderBarrierMethod = null;
  private static boolean mOrderMethodsFetched = false;

  /**
   * Enables Z support for the Canvas. The method is publicly available starting from API 29 and was
   * hidden before, so we have to resort to reflection tricks to ensure we can use this API.
   */
  @SuppressLint({"SoonBlockedPrivateApi", "PrivateApi"})
  public static void enableZ(Canvas canvas, boolean enable) {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
      return;
    }

    if (Build.VERSION.SDK_INT >= 29) {
      if (enable) {
        canvas.enableZ();
      } else {
        canvas.disableZ();
      }
    } else {
      fetchOrderMethods();
      try {
        if (enable && mReorderBarrierMethod != null) {
          mReorderBarrierMethod.invoke(canvas);
        }
        if (!enable && mInorderBarrierMethod != null) {
          mInorderBarrierMethod.invoke(canvas);
        }
      } catch (IllegalAccessException | InvocationTargetException ignore) {
        // Do nothing
      }
    }
  }

  private static void fetchOrderMethods() {
    if (!mOrderMethodsFetched) {
      try {
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.P) {
          // use double reflection to avoid grey list on P
          Method getDeclaredMethod =
              Class.class.getDeclaredMethod("getDeclaredMethod", String.class, Class[].class);
          mReorderBarrierMethod =
              (Method) getDeclaredMethod.invoke(Canvas.class, "insertReorderBarrier", new Class[0]);
          mInorderBarrierMethod =
              (Method) getDeclaredMethod.invoke(Canvas.class, "insertInorderBarrier", new Class[0]);
        } else {
          mReorderBarrierMethod = Canvas.class.getDeclaredMethod("insertReorderBarrier");
          mInorderBarrierMethod = Canvas.class.getDeclaredMethod("insertInorderBarrier");
        }

        if (mReorderBarrierMethod == null || mInorderBarrierMethod == null) {
          return;
        }

        mReorderBarrierMethod.setAccessible(true);
        mInorderBarrierMethod.setAccessible(true);
      } catch (IllegalAccessException ignore) {
        // Do nothing
      } catch (InvocationTargetException ignore) {
        // Do nothing
      } catch (NoSuchMethodException ignore) {
        // Do nothing
      }
      mOrderMethodsFetched = true;
    }
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/views/view/CanvasUtil.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.views.view

import android.annotation.SuppressLint
import android.graphics.Canvas
import android.os.Build
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

/**
 * Copied from
 * [ Compose canvas utils ](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/ui/ui-graphics/src/androidMain/kotlin/androidx/compose/ui/graphics/CanvasUtils.android.kt;drc=3b2dde134afab8d58b9c39ad4820eaf9a6e014a9)
 */
internal object CanvasUtil {
  private var reorderBarrierMethod: Method? = null
  private var inorderBarrierMethod: Method? = null
  private var orderMethodsFetched = false

  /**
   * Enables Z support for the Canvas. The method is publicly available starting from API 29 and was
   * hidden before, so we have to resort to reflection tricks to ensure we can use this API.
   */
  @SuppressLint("SoonBlockedPrivateApi", "PrivateApi")
  @JvmStatic
  fun enableZ(canvas: Canvas, enable: Boolean) {
    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
      if (enable) {
        canvas.enableZ()
      } else {
        canvas.disableZ()
      }
    } else {
      fetchOrderMethods()
      try {
        if (enable && reorderBarrierMethod != null) {
          checkNotNull(reorderBarrierMethod).invoke(canvas)
        }
        if (!enable && inorderBarrierMethod != null) {
          checkNotNull(inorderBarrierMethod).invoke(canvas)
        }
      } catch (ignore: IllegalAccessException) {
        // Do nothing
      } catch (ignore: InvocationTargetException) {
        // Do nothing
      }
    }
  }

  private fun fetchOrderMethods() {
    if (!orderMethodsFetched) {
      try {
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.P) {
          // use double reflection to avoid grey list on P
          val getDeclaredMethod =
              Class::class
                  .java
                  .getDeclaredMethod(
                      "getDeclaredMethod", String::class.java, Array<Any>::class.java)
          reorderBarrierMethod =
              getDeclaredMethod.invoke(
                  Canvas::class.java, "insertReorderBarrier", arrayOfNulls<Class<*>>(0)) as Method
          inorderBarrierMethod =
              getDeclaredMethod.invoke(
                  Canvas::class.java, "insertInorderBarrier", arrayOfNulls<Class<*>>(0)) as Method
        } else {
          reorderBarrierMethod = Canvas::class.java.getDeclaredMethod("insertReorderBarrier")
          inorderBarrierMethod = Canvas::class.java.getDeclaredMethod("insertInorderBarrier")
        }
        if (reorderBarrierMethod == null || inorderBarrierMethod == null) {
          return
        }
        reorderBarrierMethod?.isAccessible = true
        inorderBarrierMethod?.isAccessible = true
      } catch (ignore: IllegalAccessException) {
        // Do nothing
      } catch (ignore: InvocationTargetException) {
        // Do nothing
      } catch (ignore: NoSuchMethodException) {
        // Do nothing
      }
      orderMethodsFetched = true
    }
  }
}
